src/rust/engine/pantsd/src/lib.rs

```diff
@@ -212,32 +292,97 @@ pub fn probe(working_dir: &Path, metadata_dir: &Path) -> Result<u16, String> {
   }
 }
 
+/// Computes a fingerprint of the relevant options for `pantsd` (see `fingerprinted_options`).
+///
+/// This fingerprint only needs to be stable within a single version of `pantsd`, since any
+/// mismatch of the fingerprint (particularly one caused by a new version) _should_ cause a
+/// mismatch.
+///
+/// TODO: Eventually, the Python `class ProcessManager` should be replaced with the `Metadata`
+/// struct in this crate, rather than having two codepaths for the reading/writing of metadata.
+pub fn fingerprint_compute(
+  build_root: &BuildRoot,
+  options_parser: &OptionParser,
+) -> Result<Fingerprint, String> {
+  let mut hasher = Sha256::new();
+  for option in fingerprinted_options(build_root)? {
+    // TODO: As the Rust options crate expands, more of this logic should be included on
+    // `OptionParser` or on `OptionValue`.
+    match option.option_type {
```

Comment:
Could this theoretically be `#[derive(Hash)]` on `OptionType` with the parsing instead focused on creating an appropriate `OptionType` value?

Reply:
Hm, possibly. But `Hasher` does not produce a very strong hash (it's limited to creating a u64 value)... it's _possibly_ strong enough for this use case (certainly `sha256` is overkill), but it doesn't gain us much benefit to use `Hash` anyway.

Reply:
Ah, yeah, the `Hasher::finish` method isn't appropriate.

I imagine one could use the `Hash`/`Hasher` interface to get the automatic visitor and handling of things like length-prefixing (to be more resistant to ambiguity/HashDoS via things like `["ab", "c"]` vs. `["a", "bc"]`), and then call a non-`Hasher` method to extract the full digest. (One would of course have to implement `Hasher::finish` somehow to satisfy the trait, either a panic or truncating to `u64`.)